### ISSUE TITLE
Add support for end-to-end acknowledgements to `kafka` source

### ISSUE THREAD

**Issue by bruceg** [timberio/vector#7460]
The `kafka` source needs to wait for a delivery acknowledgement before marking a batch of events as committed. This is complicated by the possibility of the events being acknowledged out-of-order but commits have to happen in order as an "offset" value. Some work will need to go into tracking this high water mark internally.

Ref #7336

**Comment by bruceg** [comment]
I am finding this difficult to achieve due to lifetime issues in the `rdkafka` programming interface. I have filed issue https://github.com/fede1024/rust-rdkafka/issues/368

I see @blt has also filed a related issue (maybe the same issue?) some time ago: https://github.com/fede1024/rust-rdkafka/issues/89

**Comment by bruceg** [comment]
@lukesteensen and I chatted about this today, and discussed two possible paths forward:
1. We can continue to use the `store_offset` interface to acknowledge the read data. However, this requires changes to the interface exposed by the `rdkafka` crate. Although the maintainers of that crate have been responsive in the past, no comment has been made on the issue we raised to them. We will likely have to fork the crate to make the changes ourselves, and feed it back as a PR.
2. The `commit` and `store_offsets` functions take a [`TopicPartitionList`](https://docs.rs/rdkafka/0.26.0/rdkafka/topic_partition_list/struct.TopicPartitionList.html) structure, which we could maintain and commit batches ourselves. This will likely add a bit of code overhead on our end, but requires no changes to `rdkafka` crate to proceed.
